tools/ci_build/build.py

```diff
@@ -1110,6 +1117,25 @@ def run_training_python_frontend_e2e_tests(cwd):
     # frontend tests are to be added here:
     log.info("Running python frontend e2e tests.")
 
+    def mpirun_subprocess(mpicmd, cwd):
+        log.debug("Running mpi subprocess in '{0}'\n{1}".format(cwd or os.getcwd(), mpicmd))
+        mpi = subprocess.Popen(
+            mpicmd,
+            cwd=cwd,
+            shell=True)
+        mpi.wait()
+        if mpi.returncode != 0:
+            raise TestFailure(mpicmd, "Test failed with returncode={}.".format(mpi.returncode))
```

Comment:
how about calling mpi.check_returncode() instead of throwing a new exception type?

Reply:
seems mpi is not a CompletedProcess. It does not have check_returncode()

Reply:
In addition, I plan not let individual test to throw and fail the CI. instead it shall fail with list of failed tests.